### PR TITLE
Feature: Add mock testing to ZB

### PR DESCRIPTION
**Pull request by gloriousCode**
# PR Descriptorino
I've notice an increase in being rate limited for ZB when running tests on Travis/AppVeyor, ruining our ability to get code coverage reports. It will usually show up as a test error like the following: 
```
--- FAIL: TestGetHistoricCandles (0.20s)

    zb_test.go:841: zb GetSpotKline rawKlines is nil
``` 
See: [here](https://travis-ci.com/github/thrasher-corp/gocryptotrader/jobs/389499845) [here](https://travis-ci.com/github/thrasher-corp/gocryptotrader/jobs/389499846) and [here](https://travis-ci.com/github/thrasher-corp/gocryptotrader/jobs/391556333) for some recent examples.
This PR adds mock testing to ensure we always have happy tests. We already have rate limits in place for ZB that match their documentation.

I also improved the `GetHistoricCandles` wrapper function to handle the timespan better

## Type of change
- [✔️] New feature (non-breaking change which adds functionality)

## How has this been tested
- [✔️] go test ./... -race
- [✔️] golangci-lint run

## Checklist

- [✔️] My code follows the style guidelines of this project
- [✔️] I have performed a self-review of my own code
- [✔️] I have commented my code, particularly in hard-to-understand areas
- [✔️] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [✔️] My changes generate no new warnings
- [✔️] I have added tests that prove my fix is effective or that my feature works
- [✔️] New and existing unit tests pass locally and on Travis with my changes
- [✔️] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: exchanges/zb/zb_live_test.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
+//+build !mock_test_off
```
Comment (xtda) [inline exchanges/zb/zb_live_test.go:1]: this logic is around the wrong way?

Comment (gloriousCode) [reply exchanges/zb/zb_live_test.go]: # 👀 
Thanks!

**Review by xtda — APPROVED** [review]
although the original design requirements behind GetHistoricCandles() & GetHistoricCandlesExtended() where that GetHistoricCandles would only send one request within exchange limits 

zb suck and this is good enough 

tested approved

**Comment by gloriousCode** [comment]
> although the original design requirements behind GetHistoricCandles() & GetHistoricCandlesExtended() where that GetHistoricCandles would only send one request within exchange limits

I think that's a valid issue. I've pushed a commit where `GetHistoricCandles` only makes one request from the startTime. `GetHistoricCandlesExtended` will keep going until `end` is reached or no new data is sent

**Review by shazbert — CHANGES_REQUESTED** [review]
nit 💣

File: exchanges/zb/zb_wrapper.go [hunk 2]
```diff
@@ -795,5 +792,77 @@ func (z *ZB) GetHistoricCandles(pair currency.Pair, a asset.Item, start, end tim
...
+	ret.SortCandlesByTimestamp(false)
+	return ret, nil
+}
+
+func (z *ZB) validateCandlesRequest(p currency.Pair, a asset.Item, start, end time.Time, interval kline.Interval) (kline.Item, error) {
+	if start.Equal(end) ||
+		end.After(time.Now()) ||
+		end.Before(start) ||
+		(start.IsZero() && !end.IsZero()) {
+		return kline.Item{}, fmt.Errorf("invalid time range supplied. Start: %v End %v",
```
Comment (xtda) [inline exchanges/zb/zb_wrapper.go:854]: worth adding test coverage to match this condition?

**Review by xtda — CHANGES_REQUESTED** [review]
Thanks for making the other changes 
One last thing from me and its good to go!

**Review by xtda — APPROVED** [review]
tested with newest changes 

image

**Review by shazbert — APPROVED** [review]
tack!

File: exchanges/zb/zb_wrapper.go [hunk 3]
```diff
@@ -795,5 +792,77 @@ func (z *ZB) GetHistoricCandles(pair currency.Pair, a asset.Item, start, end tim
...
+				continue
+			}
+			if startTime.Equal(candles.Data[x].KlineTime) {
+				// no new data has been sent
+				break allKlines
+			}
+			ret.Candles = append(ret.Candles, kline.Candle{
+				Time:   candles.Data[x].KlineTime,
+				Open:   candles.Data[x].Open,
+				High:   candles.Data[x].Close,
```
Comment (MadCozBadd) [inline exchanges/zb/zb_wrapper.go:831]: High = Close

Comment (gloriousCode) [reply exchanges/zb/zb_wrapper.go]: Sweet catch! I did an audit and it occurs in a few locations! I updated all the I found in the latest commit

**Review by MadCozBadd — APPROVED** [review]
LGTM!

**Review by shazbert — APPROVED** [review]
Recent changes ack

**Review by thrasher- — CHANGES_REQUESTED** [review]
Tested both live and mock, working great. Only have one pedantic nit and then this LGTG!

File: cmd/documentation/exchanges_templates/exchanges_mock_readme.tmpl [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+var mockTests = false
+
+func TestMain(m *testing.M) {
+	cfg := config.GetConfig()
+	cfg.LoadConfig("../../testdata/configtest.json")
+	your_current_exchange_nameConfig, err := cfg.GetExchangeConfig("your_current_exchange_name")
+	if err != nil {
+		log.Fatal("your_current_exchange_name Setup() init error", err)
+	}
+    your_current_exchange_nameConfig.API.AuthenticatedSupport = true
```
[the remark this answers is not in the document]
Comment (gloriousCode) [reply cmd/documentation/exchanges_templates/exchanges_mock_readme.tmpl]: https://www.emacswiki.org/pics/static/TabsSpacesBoth.png
Fixed!

[the remark this answers is not in the document]
Comment (thrasher-) [reply cmd/documentation/exchanges_templates/exchanges_mock_readme.tmpl]: My OCD can rest easy now 😆

**Review by MadCozBadd — APPROVED** [review]
new changes LGTM

File: cmd/documentation/communications_templates/telegram.tmpl [hunk 5]
```diff
@@ -11,39 +11,39 @@ developed by Telegram Messenger LLP
...
 
 + Once the bot has started you can interact with the bot using these commands
 via Telegram:
 
 ```
-/start  		- Will authenticate your ID
+/start			- Will authenticate your ID
 /status 		- Displays the status of the bot
 /help 			- Displays current command list
 /settings 	- Displays current bot settings
```
Comment (shazbert) [inline cmd/documentation/communications_templates/telegram.tmpl:49]: 👀

**Review by shazbert — CHANGES_REQUESTED** [review]
New changes look neat! This is pretty much an ACK on changes just some super uber tiny niterinos that don't need to be changed, but here they are.

File: cmd/documentation/README.md [hunk 6]
```diff
@@ -22,7 +22,7 @@ Join our slack to discuss all things related to GoCryptoTrader! [GoCryptoTrader
 
 #### This tool allows for the generation of new documentation through templating
 
-From the `gocryptotrader/cmd/documentation/` folder, using the go command: **go run documentation.go** this will auto-generate and regenerate documentation  across the **GoCryptoTrader** code base.
+From the `gocryptotrader/cmd/documentation/` folder, using the go command: **go run documentation.go** this will auto-generate and regenerate documentation	across the **GoCryptoTrader** code base.
```
[the remark this answers is not in the document]
Comment (shazbert) [reply cmd/documentation/README.md:25]: okay this seems wrong

**Review by thrasher- — APPROVED** [review]
tACK!

**Pull request merged**